From 7c47722ba63536aa04768f697653c42998e506a1 Mon Sep 17 00:00:00 2001
From: Valentin Raevsky <valentin@compulab.co.il>
Date: Mon, 14 Apr 2025 15:28:45 +0300
Subject: [PATCH 101/102] distro: Load dtb after boot efi bbinary

Signed-off-by: Valentin Raevsky <valentin@compulab.co.il>
---
 include/config_distro_bootcmd.h | 21 +++++++++++----------
 1 file changed, 11 insertions(+), 10 deletions(-)

diff --git a/include/config_distro_bootcmd.h b/include/config_distro_bootcmd.h
index 9d2a225e7e..51cbd9e815 100644
--- a/include/config_distro_bootcmd.h
+++ b/include/config_distro_bootcmd.h
@@ -145,9 +145,11 @@
 #define BOOTENV_SHARED_EFI                                                \
 	BOOTENV_EFI_BOOTMGR                                               \
 	\
-	"boot_efi_binary="                                                \
+	"load_efi_binary="                                                \
 		"load ${devtype} ${devnum}:${distro_bootpart} "           \
-			"${kernel_addr_r} efi/boot/"BOOTEFI_NAME"; "      \
+			"${kernel_addr_r} efi/boot/"BOOTEFI_NAME"\0"      \
+	\
+	"boot_efi_binary="                                                \
 		"if fdt addr -q ${fdt_addr_r}; then "                     \
 			"bootefi ${kernel_addr_r} ${fdt_addr_r};"         \
 		"else "                                                   \
@@ -163,6 +165,12 @@
 		"setenv efi_fdtfile ${fdtfile}; "                         \
 		BOOTENV_EFI_SET_FDTFILE_FALLBACK                          \
 		BOOTENV_RUN_EXTENSION_INIT                                \
+		"if test -e ${devtype} ${devnum}:${distro_bootpart} "     \
+					"efi/boot/"BOOTEFI_NAME"; then "  \
+				"echo Found EFI removable media binary "  \
+					"efi/boot/"BOOTEFI_NAME"; "       \
+				"run load_efi_binary; "                   \
+		"fi; "                                                    \
 		"for prefix in ${efi_dtb_prefixes}; do "                  \
 			"if test -e ${devtype} "                          \
 					"${devnum}:${distro_bootpart} "   \
@@ -171,14 +179,7 @@
 				BOOTENV_RUN_EXTENSION_APPLY               \
 			"fi;"                                             \
 		"done;"                                                   \
-		"run boot_efi_bootmgr;"                                   \
-		"if test -e ${devtype} ${devnum}:${distro_bootpart} "     \
-					"efi/boot/"BOOTEFI_NAME"; then "  \
-				"echo Found EFI removable media binary "  \
-					"efi/boot/"BOOTEFI_NAME"; "       \
-				"run boot_efi_binary; "                   \
-				"echo EFI LOAD FAILED: continuing...; "   \
-		"fi; "                                                    \
+		"run boot_efi_binary; "                                   \
 		"setenv efi_fdtfile\0"
 #define SCAN_DEV_FOR_EFI "run scan_dev_for_efi;"
 #else
-- 
2.34.1

